### PR TITLE
New API endpoint for re-running an execution

### PR DESCRIPTION
**Pull request by Kami**
This pull request adds a new API endpoint for re-running an action execution.

```
POST /v1/actionexecutions/<execution id>/re_run/
```

Previously, this functionality was implemented client side in the CLI, but now this functionality is also needed by ChatOps so it finally makes sense to move it to the server. The server side logic is exactly the same was the client side logic was so the existing "execution re-run" command works in  exactly the same way as well.

Having this functionality on the server also means WebUI can now easily implement re-run as well.

### CODE REVIEW

File: st2api/st2api/controllers/v1/actionexecutions.py [hunk 1]
```diff
@@ -130,6 +173,45 @@ def get(self, id, attribute, **kwargs):
...
 
 
+class ActionExecutionReRunController(ActionExecutionsControllerMixin, ResourceController):
+    supported_filters = {}
+    exclude_fields = [
+        'result',
+        'trigger_instance'
+    ]
+
+    class ExecutionParameters(object):
```
Comment (Kami) [inline st2api/st2api/controllers/v1/actionexecutions.py]: This is a "limitation" of our existing `jsexpose` decorator - in most other places we can just use DB models, but in this case API only takes dictionary with "parameters" key so we need to create a "dummy" class for that.

File: st2api/st2api/controllers/v1/actionexecutions.py [hunk 2]
```diff
@@ -141,7 +223,9 @@ class ActionExecutionsController(ActionExecutionsControllerMixin, ResourceContro
 
     children = ActionExecutionChildrenController()
     attribute = ActionExecutionAttributeController()
+    re_run = ActionExecutionReRunController()
```
Comment (lakshmi-kannan) [inline st2api/st2api/controllers/v1/actionexecutions.py]: Should we also add rerun for convenience?

File: st2api/tests/unit/controllers/v1/test_executions_simple.py [hunk 3]
```diff
@@ -281,6 +281,47 @@ def test_post_with_st2_context_in_headers_failed(self):
...
 
+    def test_re_run_success(self):
+        # Create a new execution
+        post_resp = self._do_post(LIVE_ACTION_1)
+        self.assertEqual(post_resp.status_int, 201)
+        execution_id = self._get_actionexecution_id(post_resp)
+
+        # Re-run created execution (no parameters overrides)
+        data = {}
+        re_run_resp = self.app.post_json('/v1/actionexecutions/%s/re_run' %
```
Comment (lakshmi-kannan) [inline st2api/tests/unit/controllers/v1/test_executions_simple.py]: Can you use executions instead of actionexecutions? We should deprecate actionexecutions.

File: st2api/tests/unit/controllers/v1/test_executions_simple.py [hunk 4]
```diff
@@ -281,6 +281,47 @@ def test_post_with_st2_context_in_headers_failed(self):
...
+                (execution_id), data)
+        self.assertEqual(re_run_resp.status_int, 201)
+
+        # Re-run created execution (with parameters overrides)
+        data = {'parameters': {'a': 'val1'}}
```
Comment (lakshmi-kannan) [inline st2api/tests/unit/controllers/v1/test_executions_simple.py]: I personally like this.

File: CHANGELOG.rst [hunk 5]
```diff
@@ -18,6 +18,8 @@ in development
 * Update CLI so it displays the error at the top level when using ``run``, ``execution run`` or
   ``execution get`` when executed workflow fails. (improvement)
 * Action trigger now contains execution id as opposed to liveaction id. (bug-fix)
+* Add new API endpoint for re-running an execution (``POST /actionexecutions/<id>/re_run/``).
```
Comment (lakshmi-kannan) [inline CHANGELOG.rst]: s/actionexecutions/executions.

**Comment by lakshmi-kannan** [comment]
Overall LGTM.

**Comment by manasdk** [comment]
`POST /v1/actionexecutions/<execution id>/re_run/`

re_run is not a property of an execution. This is encoding a verb in a URL - REStful still?

`POST /v1/actionexecutions/`

with 

BODY : {'re_run': '<execution_id>'}

seems more in line - no?

File: st2api/tests/unit/controllers/v1/test_executions_simple.py (hunk 4 above)
Comment (manasdk) [inline st2api/tests/unit/controllers/v1/test_executions_simple.py]: Yup. Parameter overrides are awesome.

File: st2client/st2client/client.py [hunk 6]
```diff
@@ -73,25 +75,25 @@ def __init__(self, base_url=None, auth_url=None, api_url=None, api_version=None,
 
         # Instantiate resource managers and assign appropriate API endpoint.
         self.managers = dict()
-        self.managers['Token'] = models.ResourceManager(
+        self.managers['Token'] = ResourceManager(
```
Comment (manasdk) [inline st2client/st2client/client.py]: even you are making unnecessary changes ... you monster!

**Comment by lakshmi-kannan** [comment]
@manasdk and I had a discussion on slack. I have an alternate proposal.

Maybe we could drop the re-run and say POST /executions/id creates that execution again to be more RESTish?

File: st2api/st2api/controllers/v1/actionexecutions.py (hunk 2 above)
Comment (Kami) [inline st2api/st2api/controllers/v1/actionexecutions.py]: I think we should have one or another and not both ("There should be one-- and preferably only one --obvious way to do it." :P).

File: st2api/tests/unit/controllers/v1/test_executions_simple.py (hunk 3 above)
Comment (Kami) [inline st2api/tests/unit/controllers/v1/test_executions_simple.py]: Yeah, I personally never liked "actionsexecutions", but that's a better change.

I'm fine with changing just this one controller and change other in a separate PR.

**Comment by Kami** [comment]
@manasdk The problem with REST is that there is no one way / one standard way to do it so everyone does it a bit differently.

In this case "re-run" is an operation on a resource and usually that goes to the URI.

I'm -1 for "Maybe we could drop the re-run and say POST /executions/id creates that execution" - it's not explicit and you have no idea what that operation does if you don't check the docs.

File: st2api/tests/unit/controllers/v1/test_executions_simple.py (hunk 3 above)
Comment (lakshmi-kannan) [inline st2api/tests/unit/controllers/v1/test_executions_simple.py]: I already handled this. You can just replace this and it should work. Let's deprecate actionexecutions in another PR. 
https://github.com/StackStorm/st2/blob/master/st2api/st2api/controllers/v1/root.py#L31

**Comment by manasdk** [comment]
> In this case "re-run" is an operation on a resource and usually that goes to the URI.

What? I don't understand REST. Verbs should not be in URI but whatever makes you happy.

File: st2api/tests/unit/controllers/v1/test_executions_simple.py (hunk 3 above)
Comment (Kami) [inline st2api/tests/unit/controllers/v1/test_executions_simple.py]: Ah, I see. Alright, cool, will make that change (will also update client to use just "executions" for other operations as well).

Comment (Kami) [inline st2api/tests/unit/controllers/v1/test_executions_simple.py]: 86fc3e478e96732e410a554c66ea17ea6d9b307a

Comment (lakshmi-kannan) [inline st2api/tests/unit/controllers/v1/test_executions_simple.py]: 86fc3e4 looks good.
